source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -734,5 +744,165 @@ TransformationFlattenConditionalBranch::GetFreshIds() const {
+    result = current_block->terminator()->GetSingleWordInOperand(0);
+  }
+  return result;
+}
+
+bool TransformationFlattenConditionalBranch::OpSelectArgumentsAreRestricted(
+    opt::IRContext* ir_context) {
+  switch (ir_context->grammar().target_env()) {
+    case SPV_ENV_UNIVERSAL_1_0:
+    case SPV_ENV_UNIVERSAL_1_1:
+    case SPV_ENV_UNIVERSAL_1_2:
+    case SPV_ENV_UNIVERSAL_1_3: {
+      return true;
+    }
+    default:
+      return false;
+  }
+}
+
+void TransformationFlattenConditionalBranch::AddBooleanVectorConstructorToBlock(
+    uint32_t fresh_id, uint32_t dimension,
+    const opt::Operand& branch_condition_operand, opt::IRContext* ir_context,
+    opt::BasicBlock* block) const {
+  opt::Instruction::OperandList in_operands;
+  for (uint32_t i = 0; i < dimension; i++) {
+    in_operands.emplace_back(branch_condition_operand);
+  }
+  block->begin()->InsertBefore(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpCompositeConstruct,
+      fuzzerutil::MaybeGetVectorType(
+          ir_context, fuzzerutil::MaybeGetBoolType(ir_context), dimension),
+      fresh_id, in_operands));
+  fuzzerutil::UpdateModuleIdBound(ir_context, fresh_id);
+}
+
+void TransformationFlattenConditionalBranch::
+    RewriteOpPhiInstructionsAtConvergenceBlock(
+        const opt::BasicBlock& header_block, uint32_t convergence_block_id,
+        opt::IRContext* ir_context) const {
+  const opt::Instruction& branch_instruction = *header_block.terminator();
+
+  const opt::Operand& branch_condition_operand =
+      branch_instruction.GetInOperand(0);
+
+  // If we encounter OpPhi instructions on vector types then we may need to
+  // introduce vector versions of the selection construct's condition to use
+  // in corresponding OpSelect instructions.  These booleans track whether we
+  // need to introduce such boolean vectors.
+  bool require_2d_boolean_vector = false;
+  bool require_3d_boolean_vector = false;
+  bool require_4d_boolean_vector = false;
+
+  // Consider every OpPhi instruction at the convergence block.
+  opt::BasicBlock* convergence_block =
+      ir_context->get_instr_block(convergence_block_id);
+  convergence_block->ForEachPhiInst(
+      [this, &branch_condition_operand, branch_instruction, &header_block,
+       ir_context, &require_2d_boolean_vector, &require_3d_boolean_vector,
+       &require_4d_boolean_vector](opt::Instruction* phi_inst) {
+        assert(phi_inst->NumInOperands() == 4 &&
+               "We are going to replace an OpPhi with an OpSelect.  This "
+               "only makes sense if the block has two distinct "
+               "predecessors.");
+        // The OpPhi takes values from two distinct predecessors.  One
+        // predecessor is associated with the "true" path of the conditional
+        // we are flattening, the other with the "false" path, but these
+        // predecessors can appear in either order as operands to the OpPhi
+        // instruction.
```

Comment:
This comment could be moved down to where we use the dominator analysis. We could also mention that we achieve this by checking dominance of the true and false immediate successors of the header block.

Reply:
Done.